From a507d7e712ef937651a743203aa3a1a5d22c9bea Mon Sep 17 00:00:00 2001
From: Fang Hui <b31070@freescale.com>
Date: Fri, 1 Aug 2014 09:09:40 +0800
Subject: [PATCH 520/635] ENGR00324993 MX6SX] Verifier/Orientation: taking
 picture block.

CameraHal::takePicture() lock mLock, while CameraHal::disableMsgType is called
in callback thread(BridgeThread). Also takePicture() will PostSyncMessage to
the thread. So dead lock.

In takePicture(), make sure the CAMERA_MSG_PREVIEW_FRAME is cleared.

Signed-off-by: Fang Hui <b31070@freescale.com>
---
 mx6/libcamera/CameraHal.cpp | 11 ++++++++++-
 1 file changed, 10 insertions(+), 1 deletion(-)

diff --git a/mx6/libcamera/CameraHal.cpp b/mx6/libcamera/CameraHal.cpp
index 2a2ac6a..a101c7b 100644
--- a/mx6/libcamera/CameraHal.cpp
+++ b/mx6/libcamera/CameraHal.cpp
@@ -1,6 +1,6 @@
 /*
  * Copyright (C) 2008 The Android Open Source Project
- * Copyright (C) 2009-2012 Freescale Semiconductor, Inc.
+ * Copyright (C) 2009-2014 Freescale Semiconductor, Inc.
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
@@ -501,8 +501,17 @@ status_t CameraHal::takePicture()
 {
     FLOG_RUNTIME("takePicture");
     status_t ret = NO_ERROR;
+
+LockAndCheck:
     Mutex::Autolock lock(mLock);
 
+    if(mMsgEnabled & CAMERA_MSG_PREVIEW_FRAME) {
+        ALOGW("takePicture, wait CAMERA_MSG_PREVIEW_FRAME to clear");
+        mLock.unlock();
+        usleep(10000);
+        goto LockAndCheck;
+    }
+
     if (!previewEnabled()) {
         FLOGE("takePicture: preview not start");
         return NO_INIT;
-- 
1.8.0

